File: src/main/java/org/example/kalkulationsprogramm/repository/LieferantDokumentRepository.java
package org.example.kalkulationsprogramm.repository;

import org.example.kalkulationsprogramm.domain.LieferantDokument;
import org.example.kalkulationsprogramm.domain.LieferantDokumentTyp;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface LieferantDokumentRepository extends JpaRepository<LieferantDokument, Long> {

        // JOIN FETCH lädt Geschäftsdaten + Uploader + Lieferant in EINEM Query mit,
        // sonst gibt es N+1 wenn das DTO später darauf zugreift.
        @Query("SELECT d FROM LieferantDokument d "
                        + "LEFT JOIN FETCH d.geschaeftsdaten "
                        + "LEFT JOIN FETCH d.uploadedBy "
                        + "LEFT JOIN FETCH d.lieferant "
                        + "WHERE d.lieferant.id = :lieferantId "
                        + "ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findByLieferantIdOrderByUploadDatumDesc(@Param("lieferantId") Long lieferantId);

        @Query("SELECT d FROM LieferantDokument d "
                        + "LEFT JOIN FETCH d.geschaeftsdaten "
                        + "LEFT JOIN FETCH d.uploadedBy "
                        + "LEFT JOIN FETCH d.lieferant "
                        + "WHERE d.lieferant.id = :lieferantId AND d.typ = :typ "
                        + "ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findByLieferantIdAndTypOrderByUploadDatumDesc(@Param("lieferantId") Long lieferantId,
                        @Param("typ") LieferantDokumentTyp typ);

        @Query("SELECT d FROM LieferantDokument d "
                        + "LEFT JOIN FETCH d.geschaeftsdaten "
                        + "LEFT JOIN FETCH d.uploadedBy "
                        + "LEFT JOIN FETCH d.lieferant "
                        + "WHERE d.lieferant.id = :lieferantId AND d.typ IN :typen "
                        + "ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findByLieferantIdAndTypIn(@Param("lieferantId") Long lieferantId,
                        @Param("typen") List<LieferantDokumentTyp> typen);

        /**
         * Findet alle Dokumente die einem Projekt zugeordnet sind (über
         * projektAnteile).
         */
        @Query("SELECT DISTINCT d FROM LieferantDokument d JOIN d.projektAnteile pa WHERE pa.projekt.id = :projektId ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findByProjektId(@Param("projektId") Long projektId);

        /**
         * Findet alle Dokumente eines bestimmten Typs die einem Projekt zugeordnet
         * sind.
         */
        @Query("SELECT DISTINCT d FROM LieferantDokument d JOIN d.projektAnteile pa WHERE pa.projekt.id = :projektId AND d.typ = :typ ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findByProjektIdAndTyp(@Param("projektId") Long projektId,
                        @Param("typ") LieferantDokumentTyp typ);

        /**
         * Prüft ob ein Dokument mit diesem Dateinamen bereits für diesen Lieferanten
         * existiert.
         * Wird für Vendor Invoice Import verwendet um Duplikate zu vermeiden.
         */
        boolean existsByLieferantIdAndOriginalDateinameContaining(Long lieferantId, String filenameFragment);

        // FETCH auf geschaeftsdaten + attachment, weil das DTO-Mapping danach
        // beide Felder anfasst (sonst N+1 ueber Treffer-Liste).
        @Query("SELECT DISTINCT d FROM LieferantDokument d " +
                        "LEFT JOIN FETCH d.geschaeftsdaten g " +
                        "LEFT JOIN FETCH d.attachment a " +
                        "WHERE d.lieferant.id = :lieferantId " +
                        "AND d.typ = org.example.kalkulationsprogramm.domain.LieferantDokumentTyp.LIEFERSCHEIN " +
                        "AND (LOWER(g.dokumentNummer) LIKE LOWER(CONCAT('%', :query, '%')) " +
                        "OR LOWER(d.originalDateiname) LIKE LOWER(CONCAT('%', :query, '%')) " +
                        "OR LOWER(a.originalFilename) LIKE LOWER(CONCAT('%', :query, '%'))) " +
                        "ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> searchLieferscheine(@Param("lieferantId") Long lieferantId,
                        @Param("query") String query);

        /**
         * Findet alle Lieferscheine die nach einem bestimmten Zeitpunkt hochgeladen
         * wurden (für Benachrichtigungen auf dem PC-Desktop).
         */
        @Query("SELECT d FROM LieferantDokument d WHERE d.typ = org.example.kalkulationsprogramm.domain.LieferantDokumentTyp.LIEFERSCHEIN AND d.uploadDatum >= :since ORDER BY d.uploadDatum DESC")
        List<LieferantDokument> findRecentLieferscheine(@Param("since") java.time.LocalDateTime since);

        /**
         * Findet das LieferantDokument, das zu einem mobilen Beleg-Scan
         * automatisch angelegt wurde (Beleg-Verknuepfung). Wird vom BelegService
         * benutzt, um beim Anzeigen eines Belegs die zugehoerige Eingangsrechnungs-
         * ID zurueckzuliefern und um doppelte Erzeugung beim Re-Run der KI zu
         * verhindern.
         */
        @Query("SELECT d FROM LieferantDokument d " +
                        "LEFT JOIN FETCH d.geschaeftsdaten " +
                        "WHERE d.beleg.id = :belegId")
        java.util.Optional<LieferantDokument> findByBelegId(@Param("belegId") Long belegId);
}

File: src/main/kotlin/org/example/kalkulationsprogramm/repository/LieferantDokumentRepository.kt
package org.example.kalkulationsprogramm.repository

import java.time.LocalDateTime
import java.util.Optional
import org.example.kalkulationsprogramm.domain.LieferantDokument
import org.example.kalkulationsprogramm.domain.LieferantDokumentTyp
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface LieferantDokumentRepository : JpaRepository<LieferantDokument, Long> {
    @Query(
        "SELECT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten " +
            "LEFT JOIN FETCH d.uploadedBy " +
            "LEFT JOIN FETCH d.lieferant " +
            "WHERE d.lieferant.id = :lieferantId " +
            "ORDER BY d.uploadDatum DESC",
    )
    fun findByLieferantIdOrderByUploadDatumDesc(
        @Param("lieferantId") lieferantId: Long?,
    ): List<LieferantDokument>

    @Query(
        "SELECT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten " +
            "LEFT JOIN FETCH d.uploadedBy " +
            "LEFT JOIN FETCH d.lieferant " +
            "WHERE d.lieferant.id = :lieferantId AND d.typ = :typ " +
            "ORDER BY d.uploadDatum DESC",
    )
    fun findByLieferantIdAndTypOrderByUploadDatumDesc(
        @Param("lieferantId") lieferantId: Long?,
        @Param("typ") typ: LieferantDokumentTyp,
    ): List<LieferantDokument>

    @Query(
        "SELECT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten " +
            "LEFT JOIN FETCH d.uploadedBy " +
            "LEFT JOIN FETCH d.lieferant " +
            "WHERE d.lieferant.id = :lieferantId AND d.typ IN :typen " +
            "ORDER BY d.uploadDatum DESC",
    )
    fun findByLieferantIdAndTypIn(
        @Param("lieferantId") lieferantId: Long?,
        @Param("typen") typen: List<LieferantDokumentTyp>,
    ): List<LieferantDokument>

    @Query("SELECT DISTINCT d FROM LieferantDokument d JOIN d.projektAnteile pa WHERE pa.projekt.id = :projektId ORDER BY d.uploadDatum DESC")
    fun findByProjektId(@Param("projektId") projektId: Long?): List<LieferantDokument>

    @Query("SELECT DISTINCT d FROM LieferantDokument d JOIN d.projektAnteile pa WHERE pa.projekt.id = :projektId AND d.typ = :typ ORDER BY d.uploadDatum DESC")
    fun findByProjektIdAndTyp(
        @Param("projektId") projektId: Long?,
        @Param("typ") typ: LieferantDokumentTyp,
    ): List<LieferantDokument>

    fun existsByLieferantIdAndOriginalDateinameContaining(lieferantId: Long?, filenameFragment: String): Boolean

    @Query(
        "SELECT DISTINCT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten g " +
            "LEFT JOIN FETCH d.attachment a " +
            "WHERE d.lieferant.id = :lieferantId " +
            "AND d.typ = org.example.kalkulationsprogramm.domain.LieferantDokumentTyp.LIEFERSCHEIN " +
            "AND (LOWER(g.dokumentNummer) LIKE LOWER(CONCAT('%', :query, '%')) " +
            "OR LOWER(d.originalDateiname) LIKE LOWER(CONCAT('%', :query, '%')) " +
            "OR LOWER(a.originalFilename) LIKE LOWER(CONCAT('%', :query, '%'))) " +
            "ORDER BY d.uploadDatum DESC",
    )
    fun searchLieferscheine(
        @Param("lieferantId") lieferantId: Long?,
        @Param("query") query: String,
    ): List<LieferantDokument>

    @Query("SELECT d FROM LieferantDokument d WHERE d.typ = org.example.kalkulationsprogramm.domain.LieferantDokumentTyp.LIEFERSCHEIN AND d.uploadDatum >= :since ORDER BY d.uploadDatum DESC")
    fun findRecentLieferscheine(@Param("since") since: LocalDateTime): List<LieferantDokument>

    @Query(
        "SELECT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten " +
            "WHERE d.beleg.id = :belegId",
    )
    fun findByBelegId(@Param("belegId") belegId: Long?): Optional<LieferantDokument>

    @Query(
        "SELECT d FROM LieferantDokument d " +
            "LEFT JOIN FETCH d.geschaeftsdaten " +
            "WHERE d.attachment IS NULL " +
            "AND LOWER(d.gespeicherterDateiname) LIKE '%.xml'",
    )
    fun findMitXmlAnzeigedatei(): List<LieferantDokument>
}
